### ISSUE TITLE
Percentages instead of pixels?

### ISSUE THREAD

**Issue by homerjam** [haltu/muuri#426]
Hi, this project looks amazing (I can't believe I only just found it)!

I'm just doing a quick evaluation before I replace Isotope and have a question. I can see that it's possible to provide your own layout algorithms so maybe this is the answer but what I would like to know is...

*Can Murri create layouts with/from items using percentage units for the CSS instead of pixels?*

ie. use `%` instead of `px` in all `top/left/translate/width/height` units?

This is how I usually create layouts for responsive sites where I don't want the order of items (layout) to change but I need the items to scale with the viewport.

Maybe there is a way to hook into how the styles are applied and do the calculations there?

Thanks!
James

**Comment by niklasramo** [comment]
Have you tried it already? You should be able to define margins and dimensions with relative units (% included) and it should just work. But it's good to keep in mind that Muuri does not automatically detect when an item's dimensions/margins change so you have to manually `grid.refreshItems(ITEMS_AND_OR_ELEMENTS_THAT_NEED_UPDATE).layout();`.

**Comment by homerjam** [comment]
Yes, unfortunately, it doesn't work as expected. Despite defining dimensions with percentage units the layout is created with pixels and there are rounding issues.

Is there a way to extend the default layout algorithm or override how styles are applied?

**Comment by h0merjam** [comment]
I think what's required is a way to override `getTranslateString()` so that we can use percentage units instead of pixels. Is this something you'd consider a PR for?

**Comment by niklasramo** [comment]
> Is there a way to extend the default layout algorithm or override how styles are applied?

Yes, you can provide your own layout algorithm which can piggyback on the default algo.

```javascript
const grid = new Muuri(elem, {
  layout: function (grid, layoutId, items, width, height, callback) {
    Muuri.defaultPacker.setOptions(grid._settings.layout);
    return Grid.defaultPacker.createLayout(
      this,
      layoutId,
      items,
      width,
      height,
      (layout) => {
        // Here you can modify the layout data before it's passed on to Muuri.
        console.log(layout.styles);
        callback(layout);
      }
    );
  },
});
```

> I think what's required is a way to override getTranslateString() so that we can use percentage units instead of pixels. Is this something you'd consider a PR for?

I'm not sure 🤔 I'd first like to investigate the rounding issues you mentioned (maybe a codepen/codesandbox?) and see if we can fix those. Muuri _should_ handle the rounding issues already pretty extensively.
